api/metrics-api-jaxrs/src/main/java/org/hawkular/metrics/api/jaxrs/handler/StringHandler.java

```diff
@@ -263,6 +277,48 @@ public void addStringData(
         observable.subscribe(new ResultSetObserver(asyncResponse));
     }
 
+    @POST
+    @Path("/raw/query")
+    @ApiOperation(value = "Fetch raw data points for multiple metrics")
+    public Response findRawData(QueryRequest query) {
+        TimeRange timeRange = new TimeRange(query.getStart(), query.getEnd());
+        if (!timeRange.isValid()) {
+            return badRequest(new ApiError(timeRange.getProblem()));
+        }
+
+        int limit;
+        if (query.getLimit() == null) {
+            limit = 0;
+        } else {
+            limit = query.getLimit();
+        }
+        Order order;
+        if (query.getOrder() == null) {
+            order = Order.defaultValue(limit, timeRange.getStart(), timeRange.getEnd());
+        } else {
+            order = Order.fromText(query.getOrder());
+        }
+
+        List<MetricId<String>> metricIds = query.getIds().stream().map(id -> new MetricId<>(tenantId, STRING, id))
+                .collect(toList());
+        Observable<NamedDataPoint<String>> dataPoints = metricsService.findDataPoints(metricIds, timeRange.getStart(),
+                timeRange.getEnd(), limit, order).observeOn(Schedulers.io());
+
+        StreamingOutput stream = output -> {
+            JsonGenerator generator = mapper.getFactory().createGenerator(output, JsonEncoding.UTF8);
+            CountDownLatch latch = new CountDownLatch(1);
+            dataPoints.subscribe(new NamedDataPointObserver<>(generator, latch, STRING));
+
+            try {
+                latch.await();
```

Comment:
> This is not right. We can't make task threads block until the whole response is loaded. This not only goes against our reactive design principle, but defeats the purpose of streaming the reponse.

We do not block until the whole response is loaded. The response starts getting streamed as soon as we get data back from Cassandra. We do not realize the whole response in memory. All of our other endpoints which do *not* stream the response do realize the entire response in memory before sending it back to the client.

Without the CountDownLatch the Response object gets returned before we finish streaming which results in an NPE when we call `generator.close()` here:

```java
public void onCompleted() {
        try {
            generator.writeEndArray();
            generator.writeEndObject();
            generator.writeEndArray();
            generator.close();
            latch.countDown();
        } catch (IOException e) {
                log.warn("Error while finishing streaming data", e);
        }
    }
```